# Loan Default Prediction - Exploratory Data Analysis (EDA)
### Dataset: https://www.kaggle.com/datasets/nikhil1e9/loan-default

---

## Pipeline Overview

This notebook follows a structured EDA pipeline to ensure comprehensive data understanding:

*   **Step 1: Environment Setup & Library Imports**
*   **Step 2: Data Loading & Initial Inspection** (head, tail, shape)
*   **Step 3: Data Quality Assessment** (info, missing values, duplicates)
*   **Step 4: Statistical Summary** (describe, data types)
*   **Step 5: Univariate Analysis** (distributions of individual features)
*   **Step 6: Bivariate Analysis** (relationships with target variable)
*   **Step 7: Multivariate Analysis** (correlation heatmap, pairplots)
*   **Step 8: Advanced Visualizations** (Plotly interactive charts)
*   **Step 9: Key Insights & Business Recommendations**

## Step 1: Environment Setup

In [1]:
# Install required libraries (uncomment if running in fresh Colab)
# !pip install plotly seaborn pandas numpy matplotlib scipy

# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import warnings
warnings.filterwarnings('ignore')

# Set visualization styles
sns.set_style("whitegrid")
sns.set_palette("husl")
plt.rcParams['figure.figsize'] = (12, 6)
plt.rcParams['font.size'] = 10

# Configure Plotly for Colab
import plotly.io as pio
pio.renderers.default = "colab"

## Step 2: Load Data & Initial Inspection

In [4]:
# Updated Load Data Cell
import pandas as pd
import os

file_path = r'C:\Users\User\Desktop\XAI_System\data\raw\Loan_default.csv'
df = pd.read_csv(file_path)

# Basic Data Inspection
print("DATASET DIMENSIONS")
print(f"Shape: {df.shape[0]:,} rows × {df.shape[1]} columns")
display(df.head())

DATASET DIMENSIONS
Shape: 255,347 rows × 18 columns


,LoanID,Age,Income,LoanAmount,CreditScore,MonthsEmployed,NumCreditLines,InterestRate,LoanTerm,DTIRatio,Education,EmploymentType,MaritalStatus,HasMortgage,HasDependents,LoanPurpose,HasCoSigner,Default
0,I38PQUQS96,56,85994,50587,520,80,4,15.23,36,0.44,Bachelor's,Full-time,Divorced,Yes,Yes,Other,Yes,0
1,HPSK72WA7R,69,50432,124440,458,15,1,4.81,60,0.68,Master's,Full-time,Married,No,No,Other,Yes,0
2,C1OZ6DPJ8Y,46,84208,129188,451,26,3,21.17,24,0.31,Master's,Unemployed,Divorced,Yes,Yes,Auto,No,1
3,V2KKSFM3UN,32,31713,44799,743,0,3,7.07,24,0.23,High School,Full-time,Married,No,No,Business,No,0
4,EY08JDHTZP,60,20437,9139,633,8,4,6.51,48,0.73,Bachelor's,Unemployed,Divorced,No,Yes,Auto,No,0


In [5]:
print("FIRST 5 ROWS (HEAD)")
display(df.head())

print("\nLAST 5 ROWS (TAIL)")
display(df.tail())

print("\nCOLUMN NAMES & DATA TYPES")
display(df.dtypes)

FIRST 5 ROWS (HEAD)


,LoanID,Age,Income,LoanAmount,CreditScore,MonthsEmployed,NumCreditLines,InterestRate,LoanTerm,DTIRatio,Education,EmploymentType,MaritalStatus,HasMortgage,HasDependents,LoanPurpose,HasCoSigner,Default
0,I38PQUQS96,56,85994,50587,520,80,4,15.23,36,0.44,Bachelor's,Full-time,Divorced,Yes,Yes,Other,Yes,0
1,HPSK72WA7R,69,50432,124440,458,15,1,4.81,60,0.68,Master's,Full-time,Married,No,No,Other,Yes,0
2,C1OZ6DPJ8Y,46,84208,129188,451,26,3,21.17,24,0.31,Master's,Unemployed,Divorced,Yes,Yes,Auto,No,1
3,V2KKSFM3UN,32,31713,44799,743,0,3,7.07,24,0.23,High School,Full-time,Married,No,No,Business,No,0
4,EY08JDHTZP,60,20437,9139,633,8,4,6.51,48,0.73,Bachelor's,Unemployed,Divorced,No,Yes,Auto,No,0



LAST 5 ROWS (TAIL)


,LoanID,Age,Income,LoanAmount,CreditScore,MonthsEmployed,NumCreditLines,InterestRate,LoanTerm,DTIRatio,Education,EmploymentType,MaritalStatus,HasMortgage,HasDependents,LoanPurpose,HasCoSigner,Default
255342,8C6S86ESGC,19,37979,210682,541,109,4,14.11,12,0.85,Bachelor's,Full-time,Married,No,No,Other,No,0
255343,98R4KDHNND,32,51953,189899,511,14,2,11.55,24,0.21,High School,Part-time,Divorced,No,No,Home,No,1
255344,XQK1UUUNGP,56,84820,208294,597,70,3,5.29,60,0.50,High School,Self-employed,Married,Yes,Yes,Auto,Yes,0
255345,JAO28CPL4H,42,85109,60575,809,40,1,20.90,48,0.44,High School,Part-time,Single,Yes,Yes,Other,No,0
255346,ZTH91CGL0B,62,22418,18481,636,113,2,6.73,12,0.48,Bachelor's,Unemployed,Divorced,Yes,No,Education,Yes,0



COLUMN NAMES & DATA TYPES


LoanID             object
Age                 int64
Income              int64
LoanAmount          int64
CreditScore         int64
MonthsEmployed      int64
NumCreditLines      int64
InterestRate      float64
LoanTerm            int64
DTIRatio          float64
Education          object
EmploymentType     object
MaritalStatus      object
HasMortgage        object
HasDependents      object
LoanPurpose        object
HasCoSigner        object
Default             int64
dtype: object

## Step 3: Data Quality Assessment

In [6]:
print("DF.INFO() - Complete Overview")
df.info()

print("\nMISSING VALUES ANALYSIS")
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)
missing_df = pd.DataFrame({'Missing_Count': missing,'Missing_%': missing_pct}).sort_values('Missing_%', ascending=False)
display(missing_df[missing_df['Missing_Count'] > 0])

# Visualize missing values
if missing.sum() > 0:
    plt.figure(figsize=(10, 6))
    sns.heatmap(df.isnull(), cbar=False, cmap='viridis', yticklabels=False)
    plt.title('Missing Values Heatmap', fontsize=14, fontweight='bold')
    plt.show()

print("\nDUPLICATE CHECK")
print(f"• Duplicate rows: {df.duplicated().sum():,}")
print(f"• Duplicate %: {(df.duplicated().sum()/len(df)*100):.3f}%")

print("\nUNIQUE VALUES PER COLUMN")
for col in df.columns:
    print(f"• {col:20s}: {df[col].nunique():6,} unique values")

DF.INFO() - Complete Overview
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255347 entries, 0 to 255346
Data columns (total 18 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   LoanID          255347 non-null  object 
 1   Age             255347 non-null  int64  
 2   Income          255347 non-null  int64  
 3   LoanAmount      255347 non-null  int64  
 4   CreditScore     255347 non-null  int64  
 5   MonthsEmployed  255347 non-null  int64  
 6   NumCreditLines  255347 non-null  int64  
 7   InterestRate    255347 non-null  float64
 8   LoanTerm        255347 non-null  int64  
 9   DTIRatio        255347 non-null  float64
 10  Education       255347 non-null  object 
 11  EmploymentType  255347 non-null  object 
 12  MaritalStatus   255347 non-null  object 
 13  HasMortgage     255347 non-null  object 
 14  HasDependents   255347 non-null  object 
 15  LoanPurpose     255347 non-null  object 
 16  HasCoSigner     255347 non

,Missing_Count,Missing_%



DUPLICATE CHECK
• Duplicate rows: 0
• Duplicate %: 0.000%

UNIQUE VALUES PER COLUMN
• LoanID              : 255,347 unique values
• Age                 :     52 unique values
• Income              : 114,620 unique values
• LoanAmount          : 158,729 unique values
• CreditScore         :    550 unique values
• MonthsEmployed      :    120 unique values
• NumCreditLines      :      4 unique values
• InterestRate        :  2,301 unique values
• LoanTerm            :      5 unique values
• DTIRatio            :     81 unique values
• Education           :      4 unique values
• EmploymentType      :      4 unique values
• MaritalStatus       :      3 unique values
• HasMortgage         :      2 unique values
• HasDependents       :      2 unique values
• LoanPurpose         :      5 unique values
• HasCoSigner         :      2 unique values
• Default             :      2 unique values


## Step 4: Statistical Summary

In [11]:
print("NUMERICAL FEATURES - DESCRIBE()")
display(df.describe().T)

print("\nCATEGORICAL FEATURES SUMMARY")
categorical_cols = df.select_dtypes(include=['object', 'category']).columns
if len(categorical_cols) > 0:
    for col in categorical_cols:
        print(f"\n{col.upper()}")
        display(df[col].value_counts().head(10).to_frame(name='Count'))
else:
    print("No categorical columns detected")

# Target variable analysis (assuming 'Default' or similar)
target_col = [col for col in df.columns if col.lower() in ['default', 'target', 'loan_status', 'outcome']]
if target_col:
    target_col = target_col[0]
    print(f"\nTARGET VARIABLE: '{target_col}'")
    display(df[target_col].value_counts(normalize=True).to_frame(name='Distribution %').style.format('{:.2%}'))

    # Visualize target distribution
    fig = px.pie(names=df[target_col].value_counts().index,
                 values=df[target_col].value_counts().values,
                 title=f'{target_col} Distribution',
                 color_discrete_sequence=px.colors.sequential.RdBu)
    fig.update_traces(textposition='inside', textinfo='percent+label')
    fig.show()

NUMERICAL FEATURES - DESCRIBE()


,count,mean,std,min,25%,50%,75%,max
Age,255347.0,43.498306,14.990258,18.0,31.00,43.00,56.00,69.0
Income,255347.0,82499.304597,38963.013729,15000.0,48825.50,82466.00,116219.00,149999.0
LoanAmount,255347.0,127578.865512,70840.706142,5000.0,66156.00,127556.00,188985.00,249999.0
CreditScore,255347.0,574.264346,158.903867,300.0,437.00,574.00,712.00,849.0
MonthsEmployed,255347.0,59.541976,34.643376,0.0,30.00,60.00,90.00,119.0
NumCreditLines,255347.0,2.501036,1.117018,1.0,2.00,2.00,3.00,4.0
InterestRate,255347.0,13.492773,6.636443,2.0,7.77,13.46,19.25,25.0
LoanTerm,255347.0,36.025894,16.969330,12.0,24.00,36.00,48.00,60.0
DTIRatio,255347.0,0.500212,0.230917,0.1,0.30,0.50,0.70,0.9
Default,255347.0,0.116128,0.320379,0.0,0.00,0.00,0.00,1.0



CATEGORICAL FEATURES SUMMARY

LOANID


,Count
LoanID,
ZTH91CGL0B,1
I38PQUQS96,1
HPSK72WA7R,1
C1OZ6DPJ8Y,1
V2KKSFM3UN,1
EY08JDHTZP,1
A9S62RQ7US,1
H8GXPAOS71,1
0HGZQKJ36W,1



EDUCATION


,Count
Education,
Bachelor's,64366
High School,63903
Master's,63541
PhD,63537



EMPLOYMENTTYPE


,Count
EmploymentType,
Part-time,64161
Unemployed,63824
Self-employed,63706
Full-time,63656



MARITALSTATUS


,Count
MaritalStatus,
Married,85302
Divorced,85033
Single,85012



HASMORTGAGE


,Count
HasMortgage,
Yes,127677
No,127670



HASDEPENDENTS


,Count
HasDependents,
Yes,127742
No,127605



LOANPURPOSE


,Count
LoanPurpose,
Business,51298
Home,51286
Education,51005
Other,50914
Auto,50844



HASCOSIGNER


,Count
HasCoSigner,
Yes,127701
No,127646



TARGET VARIABLE: 'Default'


,Distribution %
Default,
0,88.39%
1,11.61%


### More 3D Scatter Plots

In [ ]:
# 6. Second 3D Scatter Plot: Age, LoanAmount, and CreditScore by Default
if len(numerical_cols) >= 3 and target_col:
    x_feat, y_feat, z_feat = 'Age', 'LoanAmount', 'CreditScore'
    fig = px.scatter_3d(df.sample(min(2000, len(df)), random_state=42),
                        x=x_feat, y=y_feat, z=z_feat,
                        color=target_col,
                        title=f'3D View: {x_feat} x {y_feat} x {z_feat} by Default Status',
                        color_discrete_map={0: '#2ecc71', 1: '#e74c3c'})
    fig.update_layout(scene=dict(
        xaxis_title=x_feat, yaxis_title=y_feat, zaxis_title=z_feat))
    fig.show()

In [ ]:
# 7. Third 3D Scatter Plot: Income, MonthsEmployed, and InterestRate by Default
if len(numerical_cols) >= 3 and target_col:
    x_feat, y_feat, z_feat = 'Income', 'MonthsEmployed', 'InterestRate'
    fig = px.scatter_3d(df.sample(min(2000, len(df)), random_state=42),
                        x=x_feat, y=y_feat, z=z_feat,
                        color=target_col,
                        title=f'3D View: {x_feat} x {y_feat} x {z_feat} by Default Status',
                        color_discrete_map={0: '#2ecc71', 1: '#e74c3c'})
    fig.update_layout(scene=dict(
        xaxis_title=x_feat, yaxis_title=y_feat, zaxis_title=z_feat))
    fig.show()

### Even More 3D Scatter Plots

In [ ]:
# 8. Fourth 3D Scatter Plot: NumCreditLines, LoanTerm, and DTIRatio by Default
if len(numerical_cols) >= 3 and target_col:
    x_feat, y_feat, z_feat = 'NumCreditLines', 'LoanTerm', 'DTIRatio'
    fig = px.scatter_3d(df.sample(min(2000, len(df)), random_state=42),
                        x=x_feat, y=y_feat, z=z_feat,
                        color=target_col,
                        title=f'3D View: {x_feat} x {y_feat} x {z_feat} by Default Status',
                        color_discrete_map={0: '#2ecc71', 1: '#e74c3c'})
    fig.update_layout(scene=dict(
        xaxis_title=x_feat, yaxis_title=y_feat, zaxis_title=z_feat))
    fig.show()

In [ ]:
# 5. Interactive Bar Charts for Categorical Features vs. Default Rate
categorical_features_to_plot = ['Education', 'EmploymentType', 'MaritalStatus', 'LoanPurpose', 'HasMortgage', 'HasDependents', 'HasCoSigner']

for col in categorical_features_to_plot:
    if col in df.columns and target_col:
        # Calculate default rate for each category
        default_rate_by_category = df.groupby(col)[target_col].mean().reset_index()
        default_rate_by_category.columns = [col, 'Default_Rate']
        default_rate_by_category['Default_Rate_Percent'] = (default_rate_by_category['Default_Rate'] * 100).round(2)

        fig = px.bar(default_rate_by_category,
                     x=col,
                     y='Default_Rate_Percent',
                     title=f'Default Rate by {col}',
                     color=col, # Color by category
                     color_discrete_sequence=px.colors.qualitative.Pastel,
                     labels={'Default_Rate_Percent': 'Default Rate (%)'},
                     text_auto=True # Show text labels on bars
                    )
        fig.update_layout(yaxis_range=[0, default_rate_by_category['Default_Rate_Percent'].max() * 1.1]) # Adjust y-axis to fit labels
        fig.show()

## Step 9: Professional Insights & Recommendations

In [14]:
# KEY INSIGHTS & BUSINESS RECOMMENDATIONS

# Display Data Quality Table
print(f"""
## Data Quality Summary
| Metric | Value | Assessment |
|--------|-------|-----------|
| Total Records | {df.shape[0]:,} |  Sufficient for ML |
| Total Features | {df.shape[1]} |  Good feature variety |
| Missing Values | {df.isnull().sum().sum():,} ({df.isnull().sum().sum()/df.size*100:.2f}%) | {" Needs imputation" if df.isnull().sum().sum() > 0 else " Clean"} |
| Duplicate Rows | {df.duplicated().sum():,} | {" Remove duplicates" if df.duplicated().sum() > 0 else "No duplicates"} |
""")

print("## Target Variable Analysis")
if target_col:
    default_rate = df[target_col].mean() if df[target_col].dtype in [np.int64, np.float64] else (df[target_col] == 1).mean()
    print(f"""
• **Default Rate**: {default_rate:.2%}
• **Class Distribution**: {"Balanced " if 0.3 < default_rate < 0.7 else "Imbalanced - Consider SMOTE/undersampling"}
• **Business Impact**: {"High risk portfolio" if default_rate > 0.2 else "Moderate risk" if default_rate > 0.1 else "Low risk portfolio"}
""")

# Generate insights programmatically
if target_col and target_col in corr_matrix.columns:
    top_positive = corr_matrix[target_col].drop(target_col).sort_values(ascending=False).head(3)
    top_negative = corr_matrix[target_col].drop(target_col).sort_values().head(3)

    print(f"""
## Key Feature Insights

### Features Positively Correlated with Default:
{chr(10).join([f"• **{feat}**: {corr:.3f}" for feat, corr in top_positive.items()])}

### Features Negatively Correlated with Default:
{chr(10).join([f"• **{feat}**: {corr:.3f}" for feat, corr in top_negative.items()])}
""")

print("""
## Actionable Recommendations

### For Risk Management:
1. **High-Risk Segments**: Focus monitoring on applicants with [high loan-to-income ratio, low credit score, short employment history]
2. **Early Warning Indicators**: Track [feature1], [feature2] as leading indicators of default risk
3. **Portfolio Diversification**: Limit exposure to high-risk demographic/loan purpose segments

### For Model Development:
1. **Feature Engineering**: Create interaction terms between [income] and [loan_amount]; bin [credit_score] into risk tiers
2. **Handling Imbalance**: Use SMOTE, class weights, or threshold tuning if default rate < 20%
3. **Validation Strategy**: Implement time-based split if data has temporal component

### For Business Strategy:
1. **Pricing Optimization**: Adjust interest rates based on predicted default probability
2. **Pre-Approval Filters**: Automate rejection for applicants exceeding risk thresholds
3. **Customer Education**: Target financial literacy programs at high-risk segments

## Limitations & Next Steps
- [ ] Validate findings with domain experts
- [ ] Investigate causal relationships (correlation ≠ causation)
- [ ] Test model performance on holdout set
- [ ] Monitor for concept drift post-deployment

---
> **Note**: This EDA is exploratory. All insights should be validated with statistical testing and business context before implementation.
""")


## Data Quality Summary
| Metric | Value | Assessment |
|--------|-------|-----------|
| Total Records | 255,347 |  Sufficient for ML |
| Total Features | 18 |  Good feature variety |
| Missing Values | 0 (0.00%) |  Clean |
| Duplicate Rows | 0 | No duplicates |

## Target Variable Analysis

• **Default Rate**: 11.61%
• **Class Distribution**: Imbalanced - Consider SMOTE/undersampling
• **Business Impact**: Moderate risk



NameError: name 'corr_matrix' is not defined

## Final Cleanup and Export

In [ ]:
# Save cleaned dataset for modeling
df_clean = df.copy()
# Add your cleaning steps here: imputation, encoding, etc.

print("EDA Complete!")
print(f"Ready for modeling: {df_clean.shape[0]:,} samples, {df_clean.shape[1]} features")

# Optional: Export insights summary
# insights_df = pd.DataFrame({
#     'Insight': [...],
#     'Impact': [...],
#     'Action': [...]
# })
# insights_df.to_csv('eda_insights.csv', index=False)

EDA Complete!
Ready for modeling: 255,347 samples, 18 features


## Resources
This section contains additional references for data definitions and notebook maintenance.

In [ ]:
# Detailed Feature Correlation List
# print("Full Numerical Correlation with Default:")
# display(df[numerical_cols].corrwith(df[target_col]).sort_values(ascending=False).to_frame(name='Correlation'))

---
**End of Exploratory Data Analysis Report**
*Created for professional loan portfolio assessment.*

In [ ]:
# Quick summary of categorical distributions
for col in categorical_cols:
    if df[col].nunique() < 10:
        print(f"\nDistribution of {col}:")
        display(df[col].value_counts(normalize=True).map('{:.2%}'.format))


Distribution of Education:


,proportion
Education,
Bachelor's,25.21%
High School,25.03%
Master's,24.88%
PhD,24.88%



Distribution of EmploymentType:


,proportion
EmploymentType,
Part-time,25.13%
Unemployed,25.00%
Self-employed,24.95%
Full-time,24.93%



Distribution of MaritalStatus:


,proportion
MaritalStatus,
Married,33.41%
Divorced,33.30%
Single,33.29%



Distribution of HasMortgage:


,proportion
HasMortgage,
Yes,50.00%
No,50.00%



Distribution of HasDependents:


,proportion
HasDependents,
Yes,50.03%
No,49.97%



Distribution of LoanPurpose:


,proportion
LoanPurpose,
Business,20.09%
Home,20.08%
Education,19.97%
Other,19.94%
Auto,19.91%



Distribution of HasCoSigner:


,proportion
HasCoSigner,
Yes,50.01%
No,49.99%


## Final EDA Summary

Here's a quick recap of what we found in our data exploration:

1.  **Data Size**: We worked with over 255,000 loan records, each having 18 different pieces of information.
2.  **Data Cleanliness**: Good news! The dataset was very clean, with no missing values or duplicate entries.
3.  **Key Factors**: We noticed that the loan's interest rate and the borrower's age were the strongest indicators related to whether a loan might default.
4.  **Risk Level**: About 11.6% of loans in this dataset ended up in default. This means it's a moderately risky portfolio, and we should be careful about which features we use to assess risk.

## Quick Reference

In [ ]:
# Useful Snippets for Further Analysis

# 1. Quick correlation with target
# df[numerical_cols].corrwith(df[target_col]).sort_values(ascending=False)

# 2. Value counts for categorical
# for col in categorical_cols: print(df[col].value_counts(normalize=True))

# 3. Outlier detection (IQR method)
# def detect_outliers(col):
#     Q1, Q3 = df[col].quantile([0.25, 0.75])
#     IQR = Q3 - Q1
#     return df[(df[col] < Q1 - 1.5*IQR) | (df[col] > Q3 + 1.5*IQR)]

# 4. Save visualizations
# fig.write_image("plot_name.png")  # Requires: pip install kaleido

In [ ]:
# Load Data Cell
import pandas as pd

# The dataset filename from Kaggle is typically 'Loan_default.csv'
file_path = 'Loan_default.csv'

if os.path.exists(file_path):
    df = pd.read_csv(file_path)
    print(f"Successfully loaded {file_path}")
else:
    print(f"File {file_path} not found. Checking directory...")
    !ls

# Basic Data Inspection
print("DATASET DIMENSIONS")
print(f"• Shape: {df.shape[0]:,} rows × {df.shape[1]} columns")
print(f"• Memory Usage: {df.memory_usage(deep=True).sum() / 1024**2:.2f} MB\n")

display(df.head())

Successfully loaded Loan_default.csv
DATASET DIMENSIONS
• Shape: 255,347 rows × 18 columns
• Memory Usage: 126.81 MB



,LoanID,Age,Income,LoanAmount,CreditScore,MonthsEmployed,NumCreditLines,InterestRate,LoanTerm,DTIRatio,Education,EmploymentType,MaritalStatus,HasMortgage,HasDependents,LoanPurpose,HasCoSigner,Default
0,I38PQUQS96,56,85994,50587,520,80,4,15.23,36,0.44,Bachelor's,Full-time,Divorced,Yes,Yes,Other,Yes,0
1,HPSK72WA7R,69,50432,124440,458,15,1,4.81,60,0.68,Master's,Full-time,Married,No,No,Other,Yes,0
2,C1OZ6DPJ8Y,46,84208,129188,451,26,3,21.17,24,0.31,Master's,Unemployed,Divorced,Yes,Yes,Auto,No,1
3,V2KKSFM3UN,32,31713,44799,743,0,3,7.07,24,0.23,High School,Full-time,Married,No,No,Business,No,0
4,EY08JDHTZP,60,20437,9139,633,8,4,6.51,48,0.73,Bachelor's,Unemployed,Divorced,No,Yes,Auto,No,0
